File: examples/go/02_medium_inject/main.go
package main

import (
	"fmt"

	"github.com/qdrant/qql/gqql"
)

func main() {
	q := "QUERY 'machine learning transformer' FROM papers LIMIT 20"

	r, _ := gqql.InjectFilter(q, "tenant_id", "=", `{"str": "acme-corp"}`)
	fmt.Println("=== String filter ===")
	fmt.Println(trunc(r, 400))

	r, _ = gqql.InjectFilter(q, "impact_factor", ">=", `{"float": 5.0}`)
	fmt.Println("\n=== Numeric filter ===")
	fmt.Println(trunc(r, 400))

	r, _ = gqql.InjectFilter(q, "is_published", "=", `{"bool": true}`)
	fmt.Println("\n=== Boolean filter ===")
	fmt.Println(trunc(r, 400))
}

func trunc(s string, n int) string {
	if len(s) > n {
		return s[:n]
	}
	return s
}

File: examples/rust/02_medium_inject/src/main.rs
use std::borrow::Cow;
use qql_core::ast::{self, Value};
use qql_core::parser::Parser;

fn main() {
    let user_query = "QUERY 'machine learning transformer' FROM papers LIMIT 20";

    // Tenant isolation — inject a string filter
    let mut stmt = Parser::parse(user_query).unwrap();
    ast::inject_filter(&mut stmt, "tenant_id", "=", &Value::Str(Cow::Borrowed("acme-corp")));
    println!("=== Tenant isolation ===");
    println!("{:#?}\n", stmt);

    // Multi-tenant + access control — inject TWO filters
    let mut stmt = Parser::parse(user_query).unwrap();
    ast::inject_filter(&mut stmt, "tenant_id", "=", &Value::Str(Cow::Borrowed("acme-corp")));
    ast::inject_filter(
        &mut stmt,
        "visibility",
        "IN",
        &Value::List(vec![
            Value::Str(Cow::Borrowed("public")),
            Value::Str(Cow::Borrowed("internal")),
        ]),
    );
    println!("=== Tenant + access control ===");
    println!("{:#?}\n", stmt);

    // Numeric filter — inject a float comparison
    let mut stmt = Parser::parse("QUERY 'covid research' FROM publications LIMIT 10").unwrap();
    ast::inject_filter(&mut stmt, "impact_factor", ">=", &Value::Float(5.0));
    println!("=== Numeric filter ===");
    println!("{:#?}\n", stmt);
}
